File: Data Structures & Algorithms/permutation-string/submission-3.py
from collections import Counter

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2): 
            return False
        
        wanted = [0] * 26
        cur = [0] * 26

        windowSize = len(s1)

        # fill wanted
        for c in s1:
            idx = ord(c) - ord('a')
            wanted[idx] += 1
        
        # fill the first windowSize chars from s2 into cur
        l = 0
        r = 0
        while r < windowSize:
            c = s2[r]
            idx = ord(c) - ord('a')
            cur[idx] += 1
            r += 1

        if wanted == cur:
                return True

        while r < len(s2):
            if wanted == cur:
                return True

            incoming = s2[r]
            outgoing = s2[l]

            cur[ord(incoming) - ord('a')] += 1
            cur[ord(outgoing) - ord('a')] -= 1

            r += 1
            l += 1

            if wanted == cur:
                return True
        
        return False
            
        

File: Data Structures & Algorithms/palindromic-substrings/submission-3.swift
class Solution {
    func countSubstrings(_ s: String) -> Int {
        let chars = Array(s)
        let n = chars.count
        
        if n < 2 {
            return n
        }

        var dp = Array(
            repeating: Array(repeating: false, count: n),
            count: n
        )

        var count = 0

        for i in 0..<n {
            dp[i][i] = true
            count += 1
        }

        for length in 2...n {
            for i in 0...(n - length) {
                let j = i + length - 1

                if chars[i] == chars[j] && (length == 2 || dp[i + 1][j - 1]) {
                    dp[i][j] = true
                    count += 1
                }
            }
        }

        return count
    }
}
